Allow choice of reading Pi and Phi from numeric ID

knelli2 opened the pull request:
Or setting them based on the 1- and 3-index constraints. Only works for numeric ID from GH vars. Numeric ID from ADM vars still computes phi with numerical derivatives.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Allow choice of reading Pi and Phi from numeric ID

Or setting them based on the 1- and 3-index constraints. Only works for
numeric ID from GH vars. Numeric ID from ADM vars still computes phi
with numerical derivatives.

nilsvu commented on support/Pipelines/Bbh/Ringdown.yaml:
```diff
@@ -19,9 +19,11 @@ InitialData:
     Variables:
       SpacetimeMetric: SpacetimeMetric
       Pi: Pi
+      Phi: Phi
```
If `SetPiAndPhiFromConstraints` is true then a Phi dataset must be present in the file but will be unused. Also, `NumericInitialData` doesn't even need to know if we want to adjust the gauge later (which is the main thing that `SetPiAndPhiFromConstraints` does in my understanding). I think here's a way to fix this:

1. Add a variant to `NumericInitialData::Variables` that contains g, Pi, and Phi (in addition to the one with only g and Pi). Then set phi=deriv(g) for the one where phi is missing, and use the numeric phi when it's given.
2. Remove the option `SetPiAndPhiFromConstraints` from the numeric ID and put it into the `const_global_cache_tags` for the `SetPiAndPhiFromConstraints` action instead, just to enable or disable the action.

knelli2 replied:
For 2, though this makes the code a bit simpler, it makes the options in the input file look ugly. The `InitialData:` in the input file isn't a group, but just a tag that holds a unique ptr to ID, so if I move `SetPiAndPhiFromConstraints` to the cache, it will just be at the root yaml level and not under `InitialData:`. It works, but isn't very organized. Any suggestions?

nilsvu replied:
Yea you can put the tag in the evolution system group, next to the gauge condition and damping functions

geoffrey4444 replied:
I prefer keeping the option in the InitialData block

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Actions/SetInitialData.hpp:
```diff
@@ -103,14 +105,16 @@ class NumericInitialData : public evolution::initial_data::InitialData {
         "numerical derivative of SpacetimeMetric, to enforce the 3-index "
-        "constraint.";
+        "constraint and Pi is set using the 1-index constraint. Otherwise, "
+        "they are read in from the numeric ID.";
```
Note that Pi is always read from numeric ID, then modified to transform to the evolution gauge if `SetPiAndPhiFromConstraints` is true

nilsdeppe reviewed (dismissed):
LGTM!!! Thanks for doing this!!

The only possible thing we may want is the ability to set only pi, only phi, both, or none. However, I think that's beyond  the scope of this PR and can be delayed until we need it. Probably the easiest way to generalize is to change the tag type from a bool to an enum with different states.

knelli2 committed:
Add tag aliases to MutateApply

Mutators run with MutateApply weren't able to add simple or cache tags,
now they are.

knelli2 commented:
Updated after discussions with @nilsvu. Decided that if we are reading GH numeric data, then we won't need to set Pi and Phi at all. Examples being PB&J and Ringdown. And if we are using ADM variables, we'll need to set Pi to work with the evolution gauge and set Phi to satisfy constraints. So I removed the option in the input file and now we just always set Pi and Phi from constraints unless we are using GH numeric ID

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Actions/SetInitialData.hpp:
```diff
@@ -96,21 +98,21 @@ class NumericInitialData : public evolution::initial_data::InitialData {
         "from these numeric fields, as well as their numeric spatial "
-        "derivatives on the computational grid.";
+        "derivatives on the computational grid. The GH variable Pi will be set "
+        "to satisfy the gauge constaint using the evolution gauge. The GH "
```
typo "constraint"
